library/std/src/sys/windows/thread_local_dtor.rs

```diff
@@ -8,10 +8,14 @@
 #[thread_local]
 static mut DESTRUCTORS: Vec<(*mut u8, unsafe extern "C" fn(*mut u8))> = Vec::new();
 
+// Ensure this can never be inlined because otherwise this may break in dylibs.
+// See #44391.
+#[inline(never)]
 pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
     DESTRUCTORS.push((t, dtor));
```

Comment:
cc @michaelwoerister @bjorn3 Could we make it so that on Windows you cannot use `#[thread_local]` from a function that may end up unable to actually reference that thread-local? See also linked issue:
*  #44391

(Though I have no idea what's happening here, since it doesn't change when/where these functions are codegenned, they're not generic and didn't have `#[inline(always)]` on it).

Reply:
> (Though I have no idea what's happening here, since it doesn't change when/where these functions are codegenned, they're not generic and didn't have #[inline(always)] on it).

Not sure either.

Reply:
Btw, the context for this PR is #94820 where I was seeing local test failures related to dtors (but this didn't seem to affect CI). However, I'm not able to reproduce it now and unfortunately back then I was rather lax in my documentation.

Reply:
Any chance LTO or ThinLTO was enabled in a way that could result in some of these definitions crossing between LLVM modules?

Alternatively, if you move the functions into different modules than the `#[thread_local] static`, could you sometimes end up with them being in separate CGUs and somehow that being enough to cause an issue?

Sadly that's also still not enough because of the whole "per-DLL/EXE" aspect.

It might be something like ThinLTO'd `librustc_driver-*.so` against `libstd-*.so` *but* the `libstd-*.so`s I'm seeing in nightly don't seem to have any LLVM bitcode sections so it can't be that.

(A custom build of `libstd-*.so` might still have LLVM bitcode - also just noticed I'm looking at Linux `.so`s not Windows `.dll`s, but I doubt that would affect whether we put LLVM bitcode in them?)

Reply:
Yeah, in general, ThinLTO can cause small functions like this one to be inlined into other modules. I haven't looked at this specific case though.